In [1]:
# import Denpendencies
import pandas as pd
import os
import matplotlib.pyplot as plt
import seaborn as sns
import time
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
from datetime import datetime

In [2]:
# Import data
# Dictionary to hold file names and their paths
file_paths = {
    'amazon1': '../Resources/Datafiniti_Amazon_Consumer_Reviews_of_Amazon_Product1.csv',
    'amazon2': '../Resources/Datafiniti_Amazon_Consumer_Reviews_of_Amazon_Product2.csv'
}

# Dictionary to hold the loaded data
data_frames = {}

# Loop through the file_paths dictionary to load each file
for file_name, relative_path in file_paths.items():
    # Construct the absolute path
    file_path = os.path.abspath(relative_path)
    
    # Attempt to load the CSV file into a DataFrame
    try:
        data_frames[file_name] = pd.read_csv(file_path,low_memory=False)
        print(f"{file_name} data loaded successfully!")
    except FileNotFoundError:
        print(f"Error: File not found at {file_path}")
    except Exception as e:
        print(f"An error occurred while loading {file_name}: {e}")


amazon1 data loaded successfully!
amazon2 data loaded successfully!


In [3]:
# Accessing data from amazon1
amazon_df1 = data_frames['amazon1']
amazon_df1

id             dateAdded           dateUpdated  \
0     AVqVGZNvQMlgsOJE6eUY  2017-03-03T16:56:05Z  2018-10-25T16:36:31Z   
1     AVqVGZNvQMlgsOJE6eUY  2017-03-03T16:56:05Z  2018-10-25T16:36:31Z   
2     AVqVGZNvQMlgsOJE6eUY  2017-03-03T16:56:05Z  2018-10-25T16:36:31Z   
3     AVqVGZNvQMlgsOJE6eUY  2017-03-03T16:56:05Z  2018-10-25T16:36:31Z   
4     AVqVGZNvQMlgsOJE6eUY  2017-03-03T16:56:05Z  2018-10-25T16:36:31Z   
...                    ...                   ...                   ...   
4995  AVqkIdZiv8e3D1O-leaJ  2017-03-06T14:59:25Z  2017-09-04T11:19:31Z   
4996  AVqkIdZiv8e3D1O-leaJ  2017-03-06T14:59:25Z  2017-09-04T11:19:31Z   
4997  AVqkIdZiv8e3D1O-leaJ  2017-03-06T14:59:25Z  2017-09-04T11:19:31Z   
4998  AVqkIdZiv8e3D1O-leaJ  2017-03-06T14:59:25Z  2017-09-04T11:19:31Z   
4999  AVqkIdZiv8e3D1O-leaJ  2017-03-06T14:59:25Z  2017-09-04T11:19:31Z   

                                                   name       asins   brand  \
0     Amazon Kindle E-Reader 6" Wifi (8th Generation...  B00ZV9PXP2  Amazon   
1     Amazon Kindle E-Reader 6" Wifi (8th Generation...  B00ZV9PXP2  Amazon   
2     Amazon Kindle E-Reader 6" Wifi (8th Generation...  B00ZV9PXP2  Amazon   
3     Amazon Kindle E-Reader 6" Wifi (8th Generation...  B00ZV9PXP2  Amazon   
4     Amazon Kindle E-Reader 6" Wifi (8th Generation...  B00ZV9PXP2  Amazon   
...                                                 ...         ...     ...   
4995  Fire Tablet with Alexa, 7" Display, 16 GB, Mag...  B018Y224PY  Amazon   
4996  Fire Tablet with Alexa, 7" Display, 16 GB, Mag...  B018Y224PY  Amazon   
4997  Fire Tablet with Alexa, 7" Display, 16 GB, Mag...  B018Y224PY  Amazon   
4998  Fire Tablet with Alexa, 7" Display, 16 GB, Mag...  B018Y224PY  Amazon   
4999  Fire Tablet with Alexa, 7" Display, 16 GB, Mag...  B018Y224PY  Amazon   

                                             categories primaryCategories  \
0     Computers,Electronics Features,Tablets,Electro...       Electronics   
1     Computers,Electronics Features,Tablets,Electro...       Electronics   
2     Computers,Electronics Features,Tablets,Electro...       Electronics   
3     Computers,Electronics Features,Tablets,Electro...       Electronics   
4     Computers,Electronics Features,Tablets,Electro...       Electronics   
...                                                 ...               ...   
4995  Tablets,Fire Tablets,Electronics,iPad & Tablet...       Electronics   
4996  Tablets,Fire Tablets,Electronics,iPad & Tablet...       Electronics   
4997  Tablets,Fire Tablets,Electronics,iPad & Tablet...       Electronics   
4998  Tablets,Fire Tablets,Electronics,iPad & Tablet...       Electronics   
4999  Tablets,Fire Tablets,Electronics,iPad & Tablet...       Electronics   

                                              imageURLs  \
0     https://pisces.bbystatic.com/image2/BestBuy_US...   
1     https://pisces.bbystatic.com/image2/BestBuy_US...   
2     https://pisces.bbystatic.com/image2/BestBuy_US...   
3     https://pisces.bbystatic.com/image2/BestBuy_US...   
4     https://pisces.bbystatic.com/image2/BestBuy_US...   
...                                                 ...   
4995  https://images-na.ssl-images-amazon.com/images...   
4996  https://images-na.ssl-images-amazon.com/images...   
4997  https://images-na.ssl-images-amazon.com/images...   
4998  https://images-na.ssl-images-amazon.com/images...   
4999  https://images-na.ssl-images-amazon.com/images...   

                                                   keys  ...  \
0     allnewkindleereaderblack6glarefreetouchscreend...  ...   
1     allnewkindleereaderblack6glarefreetouchscreend...  ...   
2     allnewkindleereaderblack6glarefreetouchscreend...  ...   
3     allnewkindleereaderblack6glarefreetouchscreend...  ...   
4     allnewkindleereaderblack6glarefreetouchscreend...  ...   
...                                                 ...  ...   
4995  841667103150,0841667103150,firetabletwithalexa...  ...   
4996  841667103150,0841667103150,firetab

In [4]:
# Accessing data from amazon1
amazon_df2 = data_frames['amazon2']
amazon_df2

id             dateAdded           dateUpdated  \
0      AVpgNzjwLJeJML43Kpxn  2015-10-30T08:59:32Z  2019-04-25T09:08:16Z   
1      AVpgNzjwLJeJML43Kpxn  2015-10-30T08:59:32Z  2019-04-25T09:08:16Z   
2      AVpgNzjwLJeJML43Kpxn  2015-10-30T08:59:32Z  2019-04-25T09:08:16Z   
3      AVpgNzjwLJeJML43Kpxn  2015-10-30T08:59:32Z  2019-04-25T09:08:16Z   
4      AVpgNzjwLJeJML43Kpxn  2015-10-30T08:59:32Z  2019-04-25T09:08:16Z   
...                     ...                   ...                   ...   
28327  AVqkIhxunnc1JgDc3kg_  2017-03-06T14:59:43Z  2019-02-23T02:49:38Z   
28328  AVqkIhxunnc1JgDc3kg_  2017-03-06T14:59:43Z  2019-02-23T02:49:38Z   
28329  AVqkIhxunnc1JgDc3kg_  2017-03-06T14:59:43Z  2019-02-23T02:49:38Z   
28330  AVqkIhxunnc1JgDc3kg_  2017-03-06T14:59:43Z  2019-02-23T02:49:38Z   
28331  AVqkIhxunnc1JgDc3kg_  2017-03-06T14:59:43Z  2019-02-23T02:49:38Z   

                                                    name  \
0      AmazonBasics AAA Performance Alkaline Batterie...   
1      AmazonBasics AAA Performance Alkaline Batterie...   
2      AmazonBasics AAA Performance Alkaline Batterie...   
3      AmazonBasics AAA Performance Alkaline Batterie...   
4      AmazonBasics AAA Performance Alkaline Batterie...   
...                                                  ...   
28327  Fire HD 8 Tablet with Alexa, 8 HD Display, 16 ...   
28328  Fire HD 8 Tablet with Alexa, 8 HD Display, 16 ...   
28329  Fire HD 8 Tablet with Alexa, 8 HD Display, 16 ...   
28330  Fire HD 8 Tablet with Alexa, 8 HD Display, 16 ...   
28331  Fire HD 8 Tablet with Alexa, 8 HD Display, 16 ...   

                       asins         brand  \
0      B00QWO9P0O,B00LH3DMUO  Amazonbasics   
1      B00QWO9P0O,B00LH3DMUO  Amazonbasics   
2      B00QWO9P0O,B00LH3DMUO  Amazonbasics   
3      B00QWO9P0O,B00LH3DMUO  Amazonbasics   
4      B00QWO9P0O,B00LH3DMUO  Amazonbasics   
...                      ...           ...   
28327             B018T075DC        Amazon   
28328             B018T075DC        Amazon   
28329             B018T075DC        Amazon   
28330             B018T075DC        Amazon   
28331             B018T075DC        Amazon   

                                              categories primaryCategories  \
0      AA,AAA,Health,Electronics,Health & Household,C...   Health & Beauty   
1      AA,AAA,Health,Electronics,Health & Household,C...   Health & Beauty   
2      AA,AAA,Health,Electronics,Health & Household,C...   Health & Beauty   
3      AA,AAA,Health,Electronics,Health & Household,C...   Health & Beauty   
4      AA,AAA,Health,Electronics,Health & Household,C...   Health & Beauty   
...                                                  ...               ...   
28327  Fire Tablets,Tablets,All Tablets,Amazon Tablet...       Electronics   
28328  Fire Tablets,Tablets,All Tablets,Amazon Tablet...       Electronics   
28329  Fire Tablets,Tablets,All Tablets,Amazon Tablet...       Electronics   
28330  Fire Tablets,Tablets,All Tablets,Amazon Tablet...       Electronics   
28331  Fire Tablets,Tablets,All Tablets,Amazon Tablet...       Electronics   

                                               imageURLs  \
0      https://images-na.ssl-images-amazon.com/images...   
1      https://images-na.ssl-images-amazon.com/images...   
2      https://images-na.ssl-images-amazon.com/images...   
3      https://images-na.ssl-images-amazon.com/images...   
4      https://images-na.ssl-images-amazon.com/images...   
...                                                  ...   
28327  https://www.upccodesearch.com/images/barcode/0...   
28328  https://www.upccodesearch.com/images/barcode/0...   
28329  https://www.upccodesearch.com/images/barcode/0...   
28330  https://www.upccodesearch.com/images/barcode/0...   
28331  https://www.upccodesearch.com/images/barcode/0...   

                                                    keys  ...  \
0      amazonbasics/hl002619,amazonbasicsaaaperforman...  ...   
1      amazonbasics/hl002619,amazonbasicsaaaperforman...  ... 

In [5]:
# Concatenate vertically
complete_data = pd.concat([amazon_df1, amazon_df2])
complete_data.head()

id             dateAdded           dateUpdated  \
0  AVqVGZNvQMlgsOJE6eUY  2017-03-03T16:56:05Z  2018-10-25T16:36:31Z   
1  AVqVGZNvQMlgsOJE6eUY  2017-03-03T16:56:05Z  2018-10-25T16:36:31Z   
2  AVqVGZNvQMlgsOJE6eUY  2017-03-03T16:56:05Z  2018-10-25T16:36:31Z   
3  AVqVGZNvQMlgsOJE6eUY  2017-03-03T16:56:05Z  2018-10-25T16:36:31Z   
4  AVqVGZNvQMlgsOJE6eUY  2017-03-03T16:56:05Z  2018-10-25T16:36:31Z   

                                                name       asins   brand  \
0  Amazon Kindle E-Reader 6" Wifi (8th Generation...  B00ZV9PXP2  Amazon   
1  Amazon Kindle E-Reader 6" Wifi (8th Generation...  B00ZV9PXP2  Amazon   
2  Amazon Kindle E-Reader 6" Wifi (8th Generation...  B00ZV9PXP2  Amazon   
3  Amazon Kindle E-Reader 6" Wifi (8th Generation...  B00ZV9PXP2  Amazon   
4  Amazon Kindle E-Reader 6" Wifi (8th Generation...  B00ZV9PXP2  Amazon   

                                          categories primaryCategories  \
0  Computers,Electronics Features,Tablets,Electro...       Electronics   
1  Computers,Electronics Features,Tablets,Electro...       Electronics   
2  Computers,Electronics Features,Tablets,Electro...       Electronics   
3  Computers,Electronics Features,Tablets,Electro...       Electronics   
4  Computers,Electronics Features,Tablets,Electro...       Electronics   

                                           imageURLs  \
0  https://pisces.bbystatic.com/image2/BestBuy_US...   
1  https://pisces.bbystatic.com/image2/BestBuy_US...   
2  https://pisces.bbystatic.com/image2/BestBuy_US...   
3  https://pisces.bbystatic.com/image2/BestBuy_US...   
4  https://pisces.bbystatic.com/image2/BestBuy_US...   

                                                keys  ... reviews.doRecommend  \
0  allnewkindleereaderblack6glarefreetouchscreend...  ...               False   
1  allnewkindleereaderblack6glarefreetouchscreend...  ...                True   
2  allnewkindleereaderblack6glarefreetouchscreend...  ...                True   
3  allnewkindleereaderblack6glarefreetouchscreend...  ...                True   
4  allnewkindleereaderblack6glarefreetouchscreend...  ...                True   

    reviews.id reviews.numHelpful reviews.rating  \
0          NaN                0.0              3   
1          NaN                0.0              5   
2          NaN                0.0              4   
3  177283626.0                3.0              5   
4          NaN                0.0              5   

                                  reviews.sourceURLs  \
0  http://reviews.bestbuy.com/3545/5442403/review...   
1  http://reviews.bestbuy.com/3545/5442403/review...   
2  https://reviews.bestbuy.com/3545/5442403/revie...   
3  https://redsky.target.com/groot-domain-api/v1/...   
4  https://reviews.bestbuy.com/3545/5442403/revie...   

                                        reviews.text  \
0  I thought it would be as big as small paper bu...   
1  This kindle is light and easy to use especiall...   
2  Didnt know how much i'd use a kindle so went f...   
3  I am 100 happy with my purchase. I caught it o...   
4  Solid entry level Kindle. Great for kids. Gift...   

                                  reviews.title  reviews.username  \
0                                     Too small            llyyue   
1  Great light reader. Easy to use at the beach            Charmi   
2                           Great for the price      johnnyjojojo   
3                                   A Great Buy           Kdperry   
4      Solid entry-level Kindle. Great for kids       Johnnyblack   

                                          sourceURLs reviews.didPurchase  
0  https://www.newegg.com/Product/Product.aspx%25...                 NaN  
1  https://www.newegg.com/Product/Product.aspx%25...                 NaN  
2  https://www.newegg.com/Product/Product.aspx%25...                 NaN  
3  https://www.newegg.com/Product/Product.aspx%25...                 NaN  
4  https://www.newegg.com/Product/Product.aspx%25...                 NaN  

[5 rows x 25 co

In [6]:
# find the shape of dataframe
print(complete_data.shape)

(33332, 25)


In [7]:
# Find the total null values in dataframe
# Finding null values in the DataFrame
null_values = complete_data.isnull().sum()

print("Null values in each column:")
print(null_values)

Null values in each column:
id                         0
dateAdded                  0
dateUpdated                0
name                       0
asins                      0
brand                      0
categories                 0
primaryCategories          0
imageURLs                  0
keys                       0
manufacturer               0
manufacturerNumber         0
reviews.date               0
reviews.dateAdded      32280
reviews.dateSeen           0
reviews.doRecommend    12246
reviews.id             33262
reviews.numHelpful     12217
reviews.rating             0
reviews.sourceURLs         0
reviews.text               0
reviews.title             13
reviews.username           6
sourceURLs                 0
reviews.didPurchase    33323
dtype: int64


There are 33332 rows and 25 columns in complete_data. reviews.dateAdded contains 32280, reviews.id contains 33262 and reviews.didPurchase contains 33323 these columns has null values close to total number of rows so the purpose of organizing and cleaning data these columns can be dropped.

In [8]:
# Delete the empty columns
cleaned_data = complete_data.drop(columns=['id','reviews.dateAdded','reviews.id','reviews.didPurchase'])

In [9]:
# Function to convert datetime string to datetime object
def convert_to_datetime(datetime_str):
    if isinstance(datetime_str, str):
        try:
            return datetime.strptime(datetime_str, '%Y-%m-%dT%H:%M:%SZ')
        except ValueError:
            return datetime_str  # Return the original value if conversion fails
    else:
        return datetime_str  # Return the datetime object unchanged

# Iterate over each column in the DataFrame
for column in cleaned_data.columns:
    # Apply the conversion function to each element in the column
    cleaned_data[column] = cleaned_data[column].apply(convert_to_datetime)

# Print the DataFrame after conversion
print(cleaned_data)

                dateAdded         dateUpdated  \
0     2017-03-03 16:56:05 2018-10-25 16:36:31   
1     2017-03-03 16:56:05 2018-10-25 16:36:31   
2     2017-03-03 16:56:05 2018-10-25 16:36:31   
3     2017-03-03 16:56:05 2018-10-25 16:36:31   
4     2017-03-03 16:56:05 2018-10-25 16:36:31   
...                   ...                 ...   
28327 2017-03-06 14:59:43 2019-02-23 02:49:38   
28328 2017-03-06 14:59:43 2019-02-23 02:49:38   
28329 2017-03-06 14:59:43 2019-02-23 02:49:38   
28330 2017-03-06 14:59:43 2019-02-23 02:49:38   
28331 2017-03-06 14:59:43 2019-02-23 02:49:38   

                                                    name       asins   brand  \
0      Amazon Kindle E-Reader 6" Wifi (8th Generation...  B00ZV9PXP2  Amazon   
1      Amazon Kindle E-Reader 6" Wifi (8th Generation...  B00ZV9PXP2  Amazon   
2      Amazon Kindle E-Reader 6" Wifi (8th Generation...  B00ZV9PXP2  Amazon   
3      Amazon Kindle E-Reader 6" Wifi (8th Generation...  B00ZV9PXP2  Amazon   
4      Amaz

In [10]:
# Number of columns in data frame after deleting empty columns
print(cleaned_data.shape)

(33332, 21)


In [11]:
# Checking data type of columns in dataframe
complete_data.dtypes

id                      object
dateAdded               object
dateUpdated             object
name                    object
asins                   object
brand                   object
categories              object
primaryCategories       object
imageURLs               object
keys                    object
manufacturer            object
manufacturerNumber      object
reviews.date            object
reviews.dateAdded       object
reviews.dateSeen        object
reviews.doRecommend     object
reviews.id             float64
reviews.numHelpful     float64
reviews.rating           int64
reviews.sourceURLs      object
reviews.text            object
reviews.title           object
reviews.username        object
sourceURLs              object
reviews.didPurchase     object
dtype: object

In [12]:
# Record the start time
start_time = time.time()

# Identify columns with missing values
categorical_columns_with_missing = cleaned_data.select_dtypes(include='object').columns[cleaned_data.select_dtypes(include='object').isnull().any()]
numerical_columns_with_missing = cleaned_data.select_dtypes(include=['int64', 'float64']).columns[cleaned_data.select_dtypes(include=['int64', 'float64']).isnull().any(axis=0)]

# Initialize n_clusters for KMeans
n_clusters = 5

# Create a copy of the DataFrame for preprocessing
df_filled_unsupervised = cleaned_data.copy()

# Fill missing numerical values with 0
for col in numerical_columns_with_missing:
    df_filled_unsupervised[col] = df_filled_unsupervised[col].fillna(0)

# Convert datetime columns to numerical features
datetime_columns = df_filled_unsupervised.select_dtypes(include='datetime64').columns
for col in datetime_columns:
    df_filled_unsupervised[col + '_year'] = df_filled_unsupervised[col].dt.year
    df_filled_unsupervised[col + '_month'] = df_filled_unsupervised[col].dt.month
    df_filled_unsupervised[col + '_day'] = df_filled_unsupervised[col].dt.day
    df_filled_unsupervised[col + '_hour'] = df_filled_unsupervised[col].dt.hour
    df_filled_unsupervised[col + '_minute'] = df_filled_unsupervised[col].dt.minute
    df_filled_unsupervised[col + '_second'] = df_filled_unsupervised[col].dt.second
    df_filled_unsupervised.drop(columns=col, inplace=True)

# Prepare DataFrame for clustering (excluding categorical columns with missing values and 'reviews.text')
df_clustering = df_filled_unsupervised.drop(columns=categorical_columns_with_missing)
df_clustering = df_clustering.drop(columns=['reviews.text'])  # Remove 'reviews.text' from clustering

# Scale only numerical columns
numerical_columns = df_clustering.select_dtypes(include=['int64', 'float64']).columns
scaler = StandardScaler()
df_clustering_scaled = scaler.fit_transform(df_clustering[numerical_columns])

# Check the minimum of the number of samples and features
min_samples_features = min(df_clustering_scaled.shape)

# Reduce the number of components if necessary
n_components = min(100, min_samples_features - 1)

# Reduce dimensionality using PCA
pca = PCA(n_components=n_components)  # Adjust the number of components as needed
df_clustering_pca = pca.fit_transform(df_clustering_scaled)

# Perform clustering on PCA-transformed data
kmeans = KMeans(n_clusters=n_clusters, random_state=42)
df_filled_unsupervised['cluster'] = kmeans.fit_predict(df_clustering_pca)

# Fill missing categorical values based on clusters
for col in categorical_columns_with_missing:
    print(f"Filling missing values for {col} using clustering-based approach")
    # Count the number of missing values before filling
    missing_before = df_filled_unsupervised[col].isnull().sum()

    # Loop over each cluster to fill missing values
    for cluster in range(n_clusters):
        cluster_df = df_filled_unsupervised.loc[df_filled_unsupervised['cluster'] == cluster, col]

        if not cluster_df.empty:
            mode_value = cluster_df.mode().iloc[0] if not cluster_df.mode().empty else df_filled_unsupervised[col].mode()[0] if df_filled_unsupervised[col].mode().size > 0 else "Missing"
            df_filled_unsupervised.loc[(df_filled_unsupervised['cluster'] == cluster) & (df_filled_unsupervised[col].isnull()), col] = mode_value

    # Count the number of missing values after filling and calculate the filled amount
    missing_after = df_filled_unsupervised[col].isnull().sum()
    total_filled = missing_before - missing_after
    print(f"Filled {total_filled} missing values for {col}")

# Optionally, remove the cluster column if it's no longer needed
df_filled_unsupervised.drop(columns=['cluster'], inplace=True)

# Verify no missing values are left
print("\nNumber of missing values after filling:", df_filled_unsupervised.isnull().sum().sum())

# Record the end time
end_time = time.time()

# Calculate the time taken
time_taken = end_time - start_time

# Convert to minutes and seconds
minutes = int(time_taken // 60)
seconds = int(time_taken % 60)

print(f"Time taken to run the code: {minutes} minutes and {seconds} seconds")


Filling missing values for reviews.doRecommend using clustering-based approach
Filled 12246 missing values for reviews.doRecommend
Filling missing values for reviews.title using clustering-based approach
Filled 13 missing values for reviews.title
Filling missing values for reviews.username using clustering-based approach
Filled 6 missing values for reviews.username

Number of missing values after filling: 0
Time taken to run the code: 0 minutes and 3 seconds


In [13]:
# Finding null values in the DataFrame
null_values = df_filled_unsupervised.isnull().sum()

print("Null values in each column:")
print(null_values)

Null values in each column:
name                   0
asins                  0
brand                  0
categories             0
primaryCategories      0
imageURLs              0
keys                   0
manufacturer           0
manufacturerNumber     0
reviews.date           0
reviews.dateSeen       0
reviews.doRecommend    0
reviews.numHelpful     0
reviews.rating         0
reviews.sourceURLs     0
reviews.text           0
reviews.title          0
reviews.username       0
sourceURLs             0
dateAdded_year         0
dateAdded_month        0
dateAdded_day          0
dateAdded_hour         0
dateAdded_minute       0
dateAdded_second       0
dateUpdated_year       0
dateUpdated_month      0
dateUpdated_day        0
dateUpdated_hour       0
dateUpdated_minute     0
dateUpdated_second     0
dtype: int64


In [15]:
from textblob import TextBlob

# Assuming 'cleaned_data' contains your cleaned DataFrame
# Create a new column 'sentiment' to store the sentiment scores
df_filled_unsupervised['sentiment'] = df_filled_unsupervised['reviews.text'].apply(lambda text: TextBlob(text).sentiment.polarity)

# Convert the continuous sentiment scores to categorical labels (e.g., positive, neutral, negative)
df_filled_unsupervised['sentiment_label'] = pd.cut(df_filled_unsupervised['sentiment'], bins=3, labels=['negative', 'neutral', 'positive'])

# Display the DataFrame with sentiment scores and labels
print(df_filled_unsupervised[['reviews.text', 'sentiment', 'sentiment_label']])

                                            reviews.text  sentiment  \
0      I thought it would be as big as small paper bu...  -0.108974   
1      This kindle is light and easy to use especiall...   0.277778   
2      Didnt know how much i'd use a kindle so went f...   0.165625   
3      I am 100 happy with my purchase. I caught it o...   0.240497   
4      Solid entry level Kindle. Great for kids. Gift...   0.468750   
...                                                  ...        ...   
28327  I got 2 of these for my 8 yr old twins. My 11 ...   0.425000   
28328  I bought this for my niece for a Christmas gif...   0.300000   
28329  Very nice for light internet browsing, keeping...   0.420000   
28330  This Tablet does absolutely everything I want!...   0.272917   
28331  At ninety dollars, the expectionations are low...   0.417755   

      sentiment_label  
0             neutral  
1             neutral  
2             neutral  
3             neutral  
4            positive  
...

For each text, TextBlob computes a polarity score between -1 and 1, where negative values indicate negative sentiment, positive values indicate positive sentiment, and values around 0 indicate neutral sentiment.

pd.cut() divides the continuous polarity scores into three bins using bins=3, and assigns the labels 'negative', 'neutral', and 'positive' to these bins.

In [16]:
# Display complete dataframe
sentiment_df = df_filled_unsupervised
sentiment_df.head()

name       asins   brand  \
0  Amazon Kindle E-Reader 6" Wifi (8th Generation...  B00ZV9PXP2  Amazon   
1  Amazon Kindle E-Reader 6" Wifi (8th Generation...  B00ZV9PXP2  Amazon   
2  Amazon Kindle E-Reader 6" Wifi (8th Generation...  B00ZV9PXP2  Amazon   
3  Amazon Kindle E-Reader 6" Wifi (8th Generation...  B00ZV9PXP2  Amazon   
4  Amazon Kindle E-Reader 6" Wifi (8th Generation...  B00ZV9PXP2  Amazon   

                                          categories primaryCategories  \
0  Computers,Electronics Features,Tablets,Electro...       Electronics   
1  Computers,Electronics Features,Tablets,Electro...       Electronics   
2  Computers,Electronics Features,Tablets,Electro...       Electronics   
3  Computers,Electronics Features,Tablets,Electro...       Electronics   
4  Computers,Electronics Features,Tablets,Electro...       Electronics   

                                           imageURLs  \
0  https://pisces.bbystatic.com/image2/BestBuy_US...   
1  https://pisces.bbystatic.com/image2/BestBuy_US...   
2  https://pisces.bbystatic.com/image2/BestBuy_US...   
3  https://pisces.bbystatic.com/image2/BestBuy_US...   
4  https://pisces.bbystatic.com/image2/BestBuy_US...   

                                                keys manufacturer  \
0  allnewkindleereaderblack6glarefreetouchscreend...       Amazon   
1  allnewkindleereaderblack6glarefreetouchscreend...       Amazon   
2  allnewkindleereaderblack6glarefreetouchscreend...       Amazon   
3  allnewkindleereaderblack6glarefreetouchscreend...       Amazon   
4  allnewkindleereaderblack6glarefreetouchscreend...       Amazon   

  manufacturerNumber              reviews.date  ... dateAdded_minute  \
0         B00ZV9PXP2  2017-09-03T00:00:00.000Z  ...               56   
1         B00ZV9PXP2  2017-06-06T00:00:00.000Z  ...               56   
2         B00ZV9PXP2  2018-04-20T00:00:00.000Z  ...               56   
3         B00ZV9PXP2  2017-11-02T17:33:31.000Z  ...               56   
4         B00ZV9PXP2  2018-04-24T00:00:00.000Z  ...               56   

  dateAdded_second  dateUpdated_year  dateUpdated_month dateUpdated_day  \
0                5              2018                 10              25   
1                5              2018                 10              25   
2                5              2018                 10              25   
3                5              2018                 10              25   
4                5              2018                 10              25   

  dateUpdated_hour dateUpdated_minute dateUpdated_second sentiment  \
0               16                 36                 31 -0.108974   
1               16                 36                 31  0.277778   
2               16                 36                 31  0.165625   
3               16                 36                 31  0.240497   
4               16                 36                 31  0.468750   

   sentiment_label  
0          neutral  
1          neutral  
2          neutral  
3          neutral  
4         positive  

[5 rows x 33 columns]

In [19]:
# Count of Sentiments
sentiment_count  = df_filled_unsupervised['sentiment_label'].value_counts()
sentiment_count

sentiment_label
positive    17738
neutral     15328
negative      266
Name: count, dtype: int64